File: reference/python-consumer-low-latency-azure/fsi_consumer.py
"""FSI C4E Reference Consumer -- Python.

Mirrors Java FsiConsumer: manual commit, Avro deserialization, handler function
pattern, Prometheus metrics, graceful shutdown.
This is a reference implementation. Copy and adapt for your application.
"""

import logging
import signal
import time

from confluent_kafka import Consumer, KafkaError, KafkaException
from confluent_kafka.schema_registry import SchemaRegistryClient
from confluent_kafka.schema_registry.avro import AvroDeserializer
from confluent_kafka.serialization import StringDeserializer, SerializationContext, MessageField
from prometheus_client import Counter, Gauge, start_http_server

log = logging.getLogger(__name__)


class FsiConsumer:
    """FSI C4E Reference Consumer for Python applications.

    Key characteristics:
      - Avro deserialization with Schema Registry (modern AvroDeserializer API)
      - Manual offset commit (at-least-once delivery guarantee)
      - Graceful shutdown via SIGTERM/SIGINT signal handlers
      - Prometheus metrics exposed via HTTP endpoint
      - Configurable record handler via callable

    Usage:
        def handle_record(key: str, value: dict):
            print(f"Received: {key} -> {value}")

        config = {
            "group.id": "corebanking-txn-processor",
            "topics": "corebanking.transactions.v1.account-transaction",
            "bootstrap.servers": "localhost:9092",
            "schema.registry.url": "http://localhost:8081",
        }
        with FsiConsumer(config, handle_record) as consumer:
            consumer.start()  # Blocks until shutdown signal
    """

    # Prometheus metrics (shared across instances, labeled by group)
    _total_consumed = Counter(
        "fsi_consumer_total_consumed",
        "Total messages consumed successfully",
        ["group"],
    )
    _total_errors = Counter(
        "fsi_consumer_total_errors",
        "Total processing errors",
        ["group"],
    )
    _last_poll_count = Gauge(
        "fsi_consumer_last_poll_count",
        "Number of records in last poll batch",
        ["group"],
    )
    _last_commit_latency_ms = Gauge(
        "fsi_consumer_last_commit_latency_ms",
        "Last commit latency in milliseconds",
        ["group"],
    )

    def __init__(self, config: dict, handler: callable):
        """Create a consumer from the standard FSI configuration pattern.

        Args:
            config: Dict containing at minimum:
              - group.id (required): Consumer group identifier
              - topics (required): Comma-separated topic names
              - bootstrap.servers: Kafka broker address (default: localhost:9092)
              - schema.registry.url: SR endpoint (default: http://localhost:8081)
              - schema.registry.basic.auth.user.info: SR credentials
              - sasl.username: Confluent Cloud API key (enables SASL_SSL)
              - sasl.password: Confluent Cloud API secret
              - client.id: Custom client identifier
              - metrics.port: Prometheus HTTP port (default: "9091")
            handler: Callable(key: str, value: dict) invoked for each record.

        Raises:
            ValueError: If group.id or topics is missing.
        """
        self._group_id = config.get("group.id")
        if not self._group_id:
            raise ValueError("group.id is required")

        topics_str = config.get("topics", "")
        if not topics_str:
            raise ValueError("topics is required (comma-separated)")
        self._topics = [t.strip() for t in topics_str.split(",") if t.strip()]

        self._handler = handler
        self._running = False

        # Schema Registry client
        sr_config = {
            "url": config.get(
                "schema.registry.url", "http://localhost:8081"
            ),
        }
        if config.get("schema.registry.basic.auth.user.info"):
            sr_config["basic.auth.user.info"] = config[
                "schema.registry.basic.auth.user.info"
            ]

        sr_client = SchemaRegistryClient(sr_config)

        # Deserializers -- modern API
        self._key_deserializer = StringDeserializer("utf_8")
        self._value_deserializer = AvroDeserializer(sr_client)

        # Build consumer config with C4E mandatory settings
        client_id = config.get(
            "client.id",
            f"fsi-{self._group_id}-consumer-python",
        )

        consumer_config = {
            "bootstrap.servers": config.get(
                "bootstrap.servers", "localhost:9092"
            ),
            "group.id": self._group_id,
            # C4E MANDATORY: Manual commit
            "enable.auto.commit": False,
            "auto.offset.reset": "earliest",
            # Performance (latency-optimized for fraud detection)
            "max.poll.interval.ms": 600000,       # 10 min — headroom for backpressure
            "fetch.min.bytes": 1,                 # return immediately with any data
            "fetch.wait.max.ms": 0,               # no broker-side batching delay
            # Rebalance strategy
            "partition.assignment.strategy": "cooperative-sticky",
            # Azure connection management
            "socket.keepalive.enable": True,      # default FALSE — Azure ILB kills idle TCP at 4min
            "connections.max.idle.ms": 180000,    # 3min — recycle before Azure ILB fires
            "reconnect.backoff.max.ms": 1000,     # cap at 1s — default 10s too slow for fraud
            # Client identification
            "client.id": client_id,
        }

        # Static group membership (eliminates rebalance on pod restart)
        instance_id = config.get("group.instance.id")
        if instance_id:
            consumer_config["group.instance.id"] = instance_id

        # Add SASL config only if credentials provided
        if config.get("sasl.username"):
            consumer_config["security.protocol"] = "SASL_SSL"
            consumer_config["sasl.mechanism"] = "PLAIN"
            consumer_config["sasl.username"] = config["sasl.username"]
            consumer_config["sasl.password"] = config.get(
                "sasl.password", ""
            )

        self._consumer = Consumer(consumer_config)

        # Start Prometheus metrics HTTP server in daemon thread
        metrics_port = int(config.get("metrics.port", "9091"))
        try:
            start_http_server(metrics_port)
            log.info("Prometheus metrics server started on port %d", metrics_port)
        except OSError:
            log.warning(
                "Metrics port %d already in use -- metrics server not started",
                metrics_port,
            )

        # Register signal handlers for graceful shutdown
        signal.signal(signal.SIGTERM, self._signal_handler)
        signal.signal(signal.SIGINT, self._signal_handler)

        log.info(
            "FSI Consumer initialized [group=%s, topics=%s, client.id=%s]",
            self._group_id,
            self._topics,
            client_id,
        )

    def _signal_handler(self, signum, frame):
        """Handle SIGTERM/SIGINT for graceful shutdown."""
        log.info("Signal %d received -- initiating shutdown", signum)
        self.shutdown()

    def start(self):
        """Start the consume loop. Blocks until shutdown signal.

        Subscribes to configured topics and polls continuously. For each
        message: deserialize key/value, invoke handler, commit offset.
        Errors in the handler are logged and skipped (at-least-once semantics).
        """
        self._running = True
        self._consumer.subscribe(self._topics)
        log.info("Consumer subscribed to: %s", self._topics)

        try:
            while self._running:
                msg = self._consumer.poll(timeout=1.0)

                if msg is None:
                    continue

                if msg.error():
                    if msg.error().code() == KafkaError._PARTITION_EOF:
                        log.debug(
                            "Reached end of partition %s-%d at offset %d",
                            msg.topic(),
                            msg.partition(),
                            msg.offset(),
                        )
                        continue
                    log.error("Consumer error: %s", msg.error())
                    self._total_errors.labels(group=self._group_id).inc()
                    continue

                # Deserialize key and value
                try:
                    key = self._key_deserializer(
                        msg.key(),
                        SerializationContext(msg.topic(), MessageField.KEY),
                    )
                    value = self._value_deserializer(
                        msg.value(),
                        SerializationContext(msg.topic(), MessageField.VALUE),
                    )
                except Exception as e:
                    self._total_errors.labels(group=self._group_id).inc()
                    log.error(
                        "Deserialization error [topic=%s, partition=%d, offset=%d]: %s",
                        msg.topic(),
                        msg.partition(),
                        msg.offset(),
                        e,
                    )
                    continue

                # Invoke handler
                try:
                    self._handler(key, value)
                    self._total_consumed.labels(group=self._group_id).inc()
                except Exception as e:
                    self._total_errors.labels(group=self._group_id).inc()
                    log.error(
                        "Error processing record [topic=%s, partition=%d, offset=%d, key=%s]: %s",
                        msg.topic(),
                        msg.partition(),
                        msg.offset(),
                        key,
                        e,
                    )
                    # Default: log and continue (at-least-once semantics)

                # Commit offset after successful processing
                try:
                    commit_start = time.time() * 1000
                    self._consumer.commit(asynchronous=False)
                    commit_latency = time.time() * 1000 - commit_start
                    self._last_commit_latency_ms.labels(
                        group=self._group_id
                    ).set(commit_latency)
                except KafkaException as e:
                    log.error("Commit failed: %s", e)

                self._last_poll_count.labels(group=self._group_id).set(1)

        finally:
            consumed = self._total_consumed.labels(
                group=self._group_id
            )._value.get()
            errors = self._total_errors.labels(
                group=self._group_id
            )._value.get()
            log.info(
                "Closing consumer. Total consumed: %d, errors: %d",
                consumed,
                errors,
            )
            self._consumer.close()

    def shutdown(self):
        """Signal the consumer to stop. Safe to call from signal handler or
        another thread."""
        log.info("Shutdown requested for group: %s", self._group_id)
        self._running = False

    def __enter__(self):
        """Support context manager protocol."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Shutdown consumer on context manager exit."""
        self.shutdown()
        return False

File: reference/java-consumer-low-latency-azure/src/main/java/org/fsi/kafka/consumer/FsiConsumer.java
package org.fsi.kafka.consumer;

import io.confluent.kafka.serializers.KafkaAvroDeserializer;
import io.confluent.kafka.serializers.KafkaAvroDeserializerConfig;
import org.apache.avro.generic.GenericRecord;
import org.apache.kafka.clients.consumer.*;
import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.errors.WakeupException;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.management.MBeanServer;
import javax.management.ObjectName;
import java.lang.management.ManagementFactory;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.BiConsumer;

/**
 * FSI C4E Reference Consumer
 *
 * Key characteristics:
 * - Avro deserialization with Schema Registry
 * - Manual offset commit (at-least-once delivery guarantee)
 * - Graceful shutdown via wakeup + shutdown hook
 * - JMX metrics for Dynatrace (records consumed, lag, commit latency)
 * - Configurable record handler via functional interface
 *
 * Usage:
 *   var consumer = new FsiConsumer(config, (key, record) -> {
 *       // Process the record
 *       processTransaction(record);
 *   });
 *   consumer.start(); // Blocks until shutdown signal
 */
public class FsiConsumer implements AutoCloseable {

    private static final Logger log = LoggerFactory.getLogger(FsiConsumer.class);

    private final KafkaConsumer<String, GenericRecord> consumer;
    private final List<String> topics;
    private final BiConsumer<String, GenericRecord> recordHandler;
    private final AtomicBoolean running = new AtomicBoolean(false);

    // Metrics
    private final AtomicLong totalConsumed = new AtomicLong(0);
    private final AtomicLong totalErrors = new AtomicLong(0);
    private final AtomicLong lastPollCount = new AtomicLong(0);
    private final AtomicLong lastCommitLatencyMs = new AtomicLong(0);

    public FsiConsumer(Map<String, String> config, BiConsumer<String, GenericRecord> handler) {
        this.topics = Arrays.asList(config.getOrDefault("topics", "").split(","));
        this.recordHandler = handler;

        if (topics.isEmpty() || topics.get(0).isEmpty()) {
            throw new IllegalArgumentException("topics is required (comma-separated)");
        }

        Properties props = buildProperties(config);
        this.consumer = new KafkaConsumer<>(props);

        registerJmxMetrics(config.getOrDefault("group.id", "unknown"));
        registerShutdownHook();

        log.info("FSI Consumer initialized for topics: {}", topics);
    }

    private Properties buildProperties(Map<String, String> config) {
        Properties props = new Properties();

        // ── Connection ──
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
                config.getOrDefault("bootstrap.servers", "kafka.fsi.internal:9092"));

        // ── Confluent Cloud auth ──
        props.put("security.protocol", "SASL_SSL");
        props.put("sasl.mechanism", "PLAIN");
        props.put("sasl.jaas.config", config.get("sasl.jaas.config"));

        // ── Consumer group ──
        String groupId = config.get("group.id");
        if (groupId == null) throw new IllegalArgumentException("group.id is required");
        props.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);

        // ── Deserialization ──
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, KafkaAvroDeserializer.class.getName());

        // ── Schema Registry ──
        props.put(KafkaAvroDeserializerConfig.SCHEMA_REGISTRY_URL_CONFIG,
                config.getOrDefault("schema.registry.url", "https://schema.fsi.internal"));
        props.put("basic.auth.credentials.source", "USER_INFO");
        props.put("basic.auth.user.info", config.get("schema.registry.basic.auth.user.info"));

        // Return GenericRecord (not specific — allows schema evolution without recompile)
        props.put(KafkaAvroDeserializerConfig.SPECIFIC_AVRO_READER_CONFIG, "false");

        // ── Offset management (C4E MANDATORY: manual commit) ──
        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");

        // ── Performance (latency-optimized for fraud detection) ──
        props.put(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, 10);          // small batches = fast per-poll completion
        props.put(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, 600000);  // 10 min — headroom for backpressure
        props.put(ConsumerConfig.FETCH_MIN_BYTES_CONFIG, 1);            // return immediately with any data
        props.put(ConsumerConfig.FETCH_MAX_WAIT_MS_CONFIG, 0);          // no broker-side batching delay

        // ── Rebalance strategy ──
        props.put(ConsumerConfig.PARTITION_ASSIGNMENT_STRATEGY_CONFIG,
                "org.apache.kafka.clients.consumer.CooperativeStickyAssignor");

        // ── Static group membership (eliminates rebalance on pod restart) ──
        String instanceId = config.get("group.instance.id");
        if (instanceId != null) {
            props.put(ConsumerConfig.GROUP_INSTANCE_ID_CONFIG, instanceId);
        }

        // ── Azure connection management ──
        props.put("socket.keepalive.enable", "true");               // OS-level TCP keepalive probes
        props.put("connections.max.idle.ms", "180000");             // 3min — before Azure ILB 4min kill
        props.put("reconnect.backoff.max.ms", "1000");              // cap at 1s — default 10s too slow

        // ── Client identification ──
        String clientId = config.getOrDefault("client.id",
                "fsi-" + groupId + "-consumer");
        props.put(ConsumerConfig.CLIENT_ID_CONFIG, clientId);

        // ── Follower fetching (read from nearest replica) ──
        props.put(ConsumerConfig.CLIENT_RACK_CONFIG,
                config.getOrDefault("client.rack", "us-east-1"));

        // ── JMX metrics ──
        props.put(ConsumerConfig.METRICS_RECORDING_LEVEL_CONFIG, "INFO");

        return props;
    }

    // Decoupled poll/process: internal queue between poll thread and worker threads.
    // The poll thread's only job is to call poll() and maintain heartbeat.
    // Worker threads handle processing and offset tracking independently.
    private final BlockingQueue<ConsumerRecord<String, GenericRecord>> workQueue =
            new LinkedBlockingQueue<>(1000);

    // Pending offsets committed asynchronously by the poll thread
    private final ConcurrentHashMap<TopicPartition, OffsetAndMetadata> pendingOffsets =
            new ConcurrentHashMap<>();

    // Backpressure flag: when true, consumer.pause() is active
    private final AtomicBoolean backpressured = new AtomicBoolean(false);

    /**
     * Start the consume loop with decoupled poll/process pattern.
     * Blocks until shutdown signal.
     *
     * The poll thread hands records off to a worker thread pool via a BlockingQueue.
     * If the queue fills (downstream backpressure), consumer.pause() keeps heartbeats
     * alive without fetching new records. consumer.resume() fires when the queue drains.
     */
    public void start() {
        running.set(true);
        consumer.subscribe(topics);
        log.info("Consumer subscribed to: {}", topics);

        // Start worker thread for processing
        Thread workerThread = new Thread(this::processLoop, "fsi-consumer-worker");
        workerThread.setDaemon(true);
        workerThread.start();

        try {
            while (running.get()) {
                ConsumerRecords<String, GenericRecord> records =
                        consumer.poll(Duration.ofMillis(100));

                lastPollCount.set(records.count());

                // Backpressure: pause/resume based on queue capacity
                if (!backpressured.get() && workQueue.remainingCapacity() < 100) {
                    consumer.pause(consumer.assignment());
                    backpressured.set(true);
                    log.warn("Backpressure detected — pausing consumer (queue remaining: {})",
                            workQueue.remainingCapacity());
                } else if (backpressured.get() && workQueue.remainingCapacity() > 500) {
                    consumer.resume(consumer.assignment());
                    backpressured.set(false);
                    log.info("Backpressure cleared — resuming consumer");
                }

                // Enqueue records for worker threads
                for (ConsumerRecord<String, GenericRecord> record : records) {
                    try {
                        workQueue.put(record); // blocks if queue full (shouldn't happen with pause)
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        break;
                    }
                }

                // Commit pending offsets asynchronously
                if (!pendingOffsets.isEmpty()) {
                    Map<TopicPartition, OffsetAndMetadata> toCommit = new HashMap<>(pendingOffsets);
                    pendingOffsets.keySet().removeAll(toCommit.keySet());

                    long commitStart = System.currentTimeMillis();
                    consumer.commitAsync(toCommit, (offsets, ex) -> {
                        lastCommitLatencyMs.set(System.currentTimeMillis() - commitStart);
                        if (ex != null) {
                            log.warn("Async commit failed, will retry next cycle: {}", ex.getMessage());
                        }
                    });
                }
            }
        } catch (WakeupException e) {
            if (running.get()) throw e; // Unexpected wakeup
            log.info("Consumer wakeup received — shutting down");
        } finally {
            // Final synchronous commit of any remaining offsets
            if (!pendingOffsets.isEmpty()) {
                consumer.commitSync(new HashMap<>(pendingOffsets), Duration.ofSeconds(10));
            }
            log.info("Closing consumer. Total consumed: {}, errors: {}",
                    totalConsumed.get(), totalErrors.get());
            consumer.close(Duration.ofSeconds(30));
        }
    }

    /**
     * Worker thread: takes records from the queue, processes them, tracks offsets.
     */
    private void processLoop() {
        while (running.get() || !workQueue.isEmpty()) {
            try {
                ConsumerRecord<String, GenericRecord> record =
                        workQueue.poll(500, TimeUnit.MILLISECONDS);
                if (record == null) continue;

                try {
                    recordHandler.accept(record.key(), record.value());
                    totalConsumed.incrementAndGet();
                } catch (Exception e) {
                    totalErrors.incrementAndGet();
                    log.error("Error processing record [topic={}, partition={}, offset={}, key={}]: {}",
                            record.topic(), record.partition(), record.offset(),
                            record.key(), e.getMessage(), e);
                    // Default: log and continue (at-least-once semantics)
                }

                // Track offset for async commit by poll thread
                pendingOffsets.put(
                        new TopicPartition(record.topic(), record.partition()),
                        new OffsetAndMetadata(record.offset() + 1));

            } catch (InterruptedException ie) {
                Thread.currentThread().interrupt();
                break;
            }
        }
        log.info("Worker thread exiting");
    }

    /**
     * Signal the consumer to stop. Safe to call from another thread.
     */
    public void shutdown() {
        log.info("Shutdown requested");
        running.set(false);
        consumer.wakeup();
    }

    @Override
    public void close() {
        shutdown();
    }

    private void registerShutdownHook() {
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            log.info("Shutdown hook triggered");
            shutdown();
        }, "fsi-consumer-shutdown"));
    }

    // ── JMX Metrics for Dynatrace ──

    private void registerJmxMetrics(String groupId) {
        try {
            MBeanServer mbs = ManagementFactory.getPlatformMBeanServer();
            ObjectName name = new ObjectName(
                    "org.fsi.kafka.consumer:type=FsiConsumer,group=" + groupId);
            mbs.registerMBean(new FsiConsumerMetrics(), name);
            log.info("Registered JMX metrics: {}", name);
        } catch (Exception e) {
            log.warn("Failed to register JMX metrics: {}", e.getMessage());
        }
    }

    public interface FsiConsumerMetricsMBean {
        long getTotalConsumed();
        long getTotalErrors();
        long getLastPollCount();
        long getLastCommitLatencyMs();
        double getErrorRate();
    }

    private class FsiConsumerMetrics implements FsiConsumerMetricsMBean {
        public long getTotalConsumed() { return totalConsumed.get(); }
        public long getTotalErrors() { return totalErrors.get(); }
        public long getLastPollCount() { return lastPollCount.get(); }
        public long getLastCommitLatencyMs() { return lastCommitLatencyMs.get(); }
        public double getErrorRate() {
            long total = totalConsumed.get() + totalErrors.get();
            return total > 0 ? (double) totalErrors.get() / total : 0.0;
        }
    }
}
